Chepulis.mikhail

MChepulis opened the pull request.

MChepulis committed:
add-lab_1

add build script
add lab_1

MChepulis committed:
Update lab_1.cpp

add same path check
add frequency positive (>=0) check
fix braces spaces

MChepulis committed:
Update lab_1.cpp

Перенес константы из деректив препроцессора в глобальные переменные

справил флаг LOG_USER на флаг LOG_ERR там, где в лог выводится информация об ошибке

исправил SIGTERM на EXIT_SECCESS

перенёс ветвление из  main  в Init

добавил функцию удаления pid-файла и её вызов при завершении работы с ошибкой

заменил NULL на nullptr

исправил отступ

исправил ошибку в названии функции "CompareExtantion"

MChepulis committed:
Update lab_1.cpp

rename extantion to extantion

MChepulis committed:
Update lab_1.cpp

проверил в словаре, правильное правописание "exstension"

MChepulis committed:
remove is_need_work  flag

remove is_need_work  flag

MChepulis committed:
remove is_need_work

удалил остатки переменной is_need_work

MChepulis committed:
Add lab_2

add lab_2:
add mq connection
add pipe connection
add shm_mem connection
add host

SWAGLORD1 commented:
Не могли бы вы обойтись без включения bits/types/siginfo_t.h? Я извиняюсь, но на моем компе этот заголовочный файл отказывается подключатся, несмотря на то, что библиотека с ним вроде как установлена. Решения из интернета так же не помогли.

MChepulis committed:
remove bits/types/siginfo.h

remove <bits/types/siginfo.h>
add <signal.h>

MChepulis commented:
убрал все bits/types/siginfo_t.h заменив их на signal.h

MChepulis committed:
delete copy C-TOR

delete copy C-TOR for singletone classes

MChepulis commented:
сделал приватным конструктор копирования для синглтон классов

KotsMikhail commented on chepulis.mikhail/lab_2/host/ClientHandler.h:
```diff
@@ -0,0 +1,46 @@
+private:
+    sem_t* semaphore_host;
+    sem_t* semaphore;
+    sem_t* semaphore_client;
```
Эти семафоры нигде не закрываются

KotsMikhail commented on chepulis.mikhail/lab_2/host/ClientHandler.h:
```diff
@@ -0,0 +1,46 @@
+
+private:
+    sem_t* semaphore_host;
+    sem_t* semaphore;
```
Этот семафор, кажется, нигде не используется

MChepulis replied:
удалил этот семафор, т.к. сейчас он, действительно, уже нигде не используется

KotsMikhail commented on chepulis.mikhail/lab_2/host/Oracle.h:
```diff
@@ -0,0 +1,39 @@
+#include "semaphore.h"
+#include "../interfaces/Conn.h"
+
+class Oracle {
```
Чем различаются роли этого класса и класса ClientHandler? И почему он у вас синглтон, у вас же может быть несколько клиентов в системе и у вас родственная конфигурация

KotsMikhail commented on chepulis.mikhail/lab_2/host/Server.cpp:
```diff
@@ -0,0 +1,206 @@
+    }
+
+
+    std::string new_date = "25.11.2019";
```
Зачем нужна эта константа? Она вроде дальше нигде не используется

MChepulis replied:
остатки отладочной информации
убрал инициализацию new_date в строке 97

KotsMikhail commented on chepulis.mikhail/lab_2/host/Server.h:
```diff
@@ -0,0 +1,32 @@
+    void Start();
+private:
+    Server();
+    Server(Server& other);
```
Оператор копирования тоже следует приватить

MChepulis replied:
сделал оператор присваивания тоже приватным

KotsMikhail reviewed (changes requested):
Нет проверки на реальность даты. Дата 99.99.1999 воспринимается как нормальная.
В задании указано, что работу с клиентами следует производить в отдельных потоках. Потоки и процессы - это разные вещи

MChepulis committed:
Fix remarks of lab2

Подправил отступы
ClientHandler теперь не синглтон
теперь сервер создаёт потоки для общения с клиентами, а не процессы
Дабавил проверку на реальность даты
Убрал неиспользуемый семафор

MChepulis commented:
Добавил проверку реальности вводимой даты
добавил удаление semaphore_host при неудачном создании semaphore_client
Подправил отступы
ClientHandler теперь не синглтон
теперь сервер создаёт потоки для общения с клиентами, а не процессы

KotsMikhail commented on chepulis.mikhail/lab_2/host/Server.cpp:
```diff
@@ -0,0 +1,222 @@
+    return nullptr;
+}
+
+void Server::Start() {
```
Эта функция получилась слишком объемной. Нужно разнести инициализацию и основной цикл. Возможно, перенести инициализацию клиента в конструктор клиента

KotsMikhail commented on chepulis.mikhail/lab_2/host/ClientHandler.cpp:
```diff
@@ -0,0 +1,197 @@
+            if (s == -1) {
+                //TODO "убить" себя и клиента
+                //std::cout << "KILL KILL KILL KILL KILL KILL KILL" << std::endl;
+                syslog(LOG_NOTICE, "KILL KILL KILL KILL KILL KILL KILL");
```
Я заметил, убирайте)

MChepulis replied:
убрал агрессивные надписи в функции ClientHandler::Start()
они были частью отладочной информации

KotsMikhail reviewed (changes requested):
Я все ещё не понимаю, почему Oracle и ClientHandler должны быть разными классами. Они реализуют разные части одного и того же объекта и при этом, похоже, сильно пересекаются. В чем различаются их функции и должны ли эти функции быть разнесены по разным классам?

MChepulis committed:
Fix remarks2 of lab2

убрал ненужный вывод в лог файл
разбил функцию Server::Start() на 4 подфункций
1)ReadClientCount() - считывает с клавиатуры кол-во клиентов
2) void CreateClient(ClientHandler *handler)
3) void StartHandlerActivity(ClientHandler *handler)
4) void Server::Run()

добавил адекватную реакцию Сервера на то, что какой-то из( потоков взаимодействия)/клиентов не смог открыться

MChepulis commented:
убрал ненужный вывод в лог файл
разбил функцию Server::Start() на 4 подфункций
1) void ReadClientCount() - считывает с клавиатуры кол-во клиентов
2) void CreateClient(ClientHandler *handler) - создаёт новый процесс, в котором начинает своб работу клиент
3) void StartHandlerActivity(ClientHandler *handler) - создаёт новый поток, в котором начинает своб работу ClientHandler
4) void Server::Run() - основной цикл работы сервера. Происходит считывание информации от пользователя и её перенаправление в потоки

По поводу ClientHandler и Oracle:
ClientHandler находится на стороне сервера, получает некую информацию от него и отправляет её клиенту и ожидает ответа.При получении выводит результат или же по истечении таймаута закрывается. Если закрывается ClientHandler, то он должен закрыть клиента, за которого отвечает. Также ClientHandler является владельцем всех каналов межпроцессорного взаимодействия и объектов их синхронизации.

Oracle находится на стороне клиента, коим по сути и является.
В его задачи входит открывать уже созданные соединения, и закрывать (не уничтожая) их.
Oracle получает сообщения, обрабатывает их и отправляет результат

KotsMikhail commented on chepulis.mikhail/lab_2/host/my_functions.h:
```diff
@@ -0,0 +1,30 @@
+bool MessFromDate(std::string date, Message &mes);
+bool IsRealDate(int day, int month, int year);
+template <typename T>
+std::string toString(T val) {
```
Чем вас std::to_string не устраивает?

MChepulis replied:
действительно, забыл про этот болей просто вариант, заменил на него

KotsMikhail commented on chepulis.mikhail/lab_2/host/Server.cpp:
```diff
@@ -0,0 +1,237 @@
+}
+
+
+void* routine(void * arg) {
```
В реализации модуля не должно быть посторонних функций, если вам нужна вспомогательная функция - сделайте её приватной функцией класса

MChepulis replied:
перенёс все "сторонние функции" в приватные методы класса

KotsMikhail reviewed (changes requested):
Окей, насчет ClientHandler и Oracle вроде понятно, но зачем вы передаете сообщения между основным потоком и дочерними через pipe? Они же в одном процессе и могут обращаться к одним и тем же переменным. Их нужно только синхронизировать, иначе они могут обращаться к разным значениям одной и той же переменной. К тому же, в этом задании можно всё упростить ещё сильнее и создавать потоки при каждом обращении к клиентам. Принципиальный момент в том, чтобы обмен сообщениями с клиентами был асинхронным, поддерживать потоки на все время сеанса на самом деле не обязательно. Это не лучшее решение, так как создание потоков - затратная операция, но в данной задаче это приемлемое решение.

MChepulis committed:
Fix remarks3 in lab2

перенёс все "сторонние функции" в приватные методы класса
убрал функцию ToString, заменив её на std::to_string

MChepulis commented:
перенёс все "сторонние функции" в приватные методы класса
убрал функцию ToString, заменив её на std::to_string

для общения сервера и хендлера использую pipe, по ряду причин:
1) он не требует синхронизации, потому как сервер только пишет в конец, а хендлер только читает с начала
2) позволяет хендлеру не пропускать сообщения
Так, если сервер посылает сообщения со скоростью большей, чем обрабатывает хендлер, то они просто собираются в очередь  и хендлер обработает их позже, но точно обработает (столько, сколько позволит система).
Это означает, что хендлер обработает все запросы от сервера
т.к. мы вводим даты вручную, то переполнить pipe весьма проблематично
3) все хендлеры обработают каждый запрос только 1 раз, т.к. к ним в канал отправляется 1 сообщение
Значит не нужно отдельно хранить и синхронизировать то, какие хендлеры уже считали (для предотвращения повторного считывания уже обработанных данных)
Не нужно реализовывать собственную очередь заданий для хендлеров и синхронизировать её (т.к. тут уже и у сервер удлиняет, хендлер укорачивает очередь -> нужно синхронизировать)

KotsMikhail reviewed (approved):
Окей, в принципе, я засчитаю такое решение, но на будущее имейте в виду - чем меньше синхронизации между потоками - тем лучше. И тем более не стоит использовать в многопоточных программах инструменты для взаимодействия процессов. В частности, pipe буферизует все, что через него пробрасывают, и не дает никаких гарантий на время доставки. Если передавать таким образом большие сообщения, то все может работать медленно. Для таких задач существуют специальные коллекции, например, thread-safe очереди, которые мы будем разбирать в следующей лабе.
